Fix unicode display: CJK, powerline, tmux

tibotiber opened the pull request:
Hi! This is a follow up PR from #1376, taking 2e58e56 into consideration. I have isolated the issue with braille characters for `vtop` by decorating the corresponding nodes with a `braille-node` class additionally to the existing `unicode-node` class. It is ready to merge to the best of my knowledge.
Cc @rauchg 

**WIP summary:**
- [x] powerline wide unicode glyphs display
- [x] CJK chars display, `wc-node` css
- [x] multiple rendering of CJK in some cases, see  https://github.com/zeit/hyper/pull/1376#issuecomment-271876444 / https://github.com/zeit/hyper/pull/1376#issuecomment-272191734
- [x] tmux empty prompt width, see screenshot below:
![tmux](https://cloud.githubusercontent.com/assets/5635553/23095605/7c65ff7a-f647-11e6-8737-b227c0842cbb.png)

albinekb commented:
Wow, this is awesome! Will try it tonight 👍

tibotiber commented:
Screenshots of the result here: vtop and unicode + tmux
![vtop](https://cloud.githubusercontent.com/assets/5635553/23062348/5f99ff7a-f540-11e6-9e13-55c6ac65e8d0.png)
![unicode](https://cloud.githubusercontent.com/assets/5635553/23063050/917a0bc2-f543-11e6-90ff-7caf6afcc6d6.png)

rauchg commented:
Only issue is: shouldn't we respect the grid size for emojis? Both Terminal.app and iTerm seem to overlap them:

![image](https://cloud.githubusercontent.com/assets/13041/23089655/9e011f8e-f56b-11e6-917e-8c2e6b8cce66.png)

![image](https://cloud.githubusercontent.com/assets/13041/23089659/a626626e-f56b-11e6-8fbb-366172d7a173.png)

I wonder if by taking space _wider than a character_ we would be breaking some UIs ?

rauchg commented:
Also, since other Terminals limit the width of Emojis, I worry that we'd be displaying them _not as nicely_ in Hyper, because they'd essentially have an extra space

tibotiber commented:
@rauchg: great point about emojis, one important breaking UI is the cursor position. However we need to address at least powerline unicode characters (e.g. https://github.com/zeit/hyper/issues/1336#issuecomment-270670882), which is handled properly by iTerm as well. Emojis being spread over codepoints, it would be easier to isolate the [powerline glyphs](https://apw-bash-settings.readthedocs.io/en/latest/fontpatching.html#glyph-table) and give them more space. We can choose to apply flexible or double spacing, I would suggest double to keep the cursor position in check. From what I can see, this is the way iTerm handles it too.

After that, I think I should remove css decoration for braille, since it will fall in the normal unicode styling (?).

_Sidenote:_ This is getting interesting, powerline glyphs are stored in the [Unicode Private Use Area](https://en.wikipedia.org/wiki/Private_Use_Areas#Usage), along with medieval scripts, Klingon, and various languages of the Middle-earth... Should we display that with fix width or double width like powerline glyphs? 😂

rauchg commented:
@tibotiber that's a very good question. I think ultimately it would depend on each glyph in the private use area whether they want double or single width.

Let's do this: for now, let's give *all* private use area double (which would encompass Powerline glyphs), and then if a request comes up, we can think about adding a config flag for specifying _ranges_ that you want to be double width

tibotiber commented:
OK, I'll push changes today (it's Saturday here already). I'll remove the braille part then.

tibotiber commented:
_Update:_ The issue with powerline glyphs is not a width issue. I couldn't get this to work in a way that supports basic powerline glyphs as well as user defined ones, so I checked iTerm a bit more and realised they do not treat powerline glyphs (new or old) as wide chars. What happens is wide unicode simply overlaps on the next char. See screenshots for iTerm and Hyper:
![hyper](https://cloud.githubusercontent.com/assets/5635553/23095438/d688f74e-f644-11e6-9250-2160394f8168.png)
<img width="654" alt="iterm" src="https://cloud.githubusercontent.com/assets/5635553/23095439/d732afc8-f644-11e6-86f7-174296539e6c.png">
If you look at the `cat` above, both terminals treat the glyphs exactly in the same way. Now, the difference is in the powerline prompt line where there is a background color. See the branch sign being clipped in Hyper? Because Hyper encapsulates each unicode char in its own `<span />`, the wide unicode do not visibly overlap on the char to its right, since that char's background color is on top of whatever overlaps. This is not an issue in iTerm where unicode is probably not encapsulated like this. Why do we encapsulate unicode in individual containers exactly?

I removed my previous changes and only left the low hanging fruit: fixing of CJK wide chars. Still trying to find a way around the powerline issue. Summary of issues i'm trying to address and the ones fixed:

- [x] CJK chars display
- [ ] powerline wide unicode glyphs display
- [ ] tmux empty prompt width, see screenshot below:
![tmux](https://cloud.githubusercontent.com/assets/5635553/23095605/7c65ff7a-f647-11e6-8737-b227c0842cbb.png)

rauchg commented:
> Why do we encapsulate unicode in individual containers exactly?

As explained above, without the `<span>` with a fixed width, glyphs that are not included in a monospace font (such as braille which uses the "Apple Braille" font would show up in a different width (and this width varies per-character) and break rendering)

> What happens is wide unicode simply overflows on the next char

That's not true for Emojis it seems? Or emojis are not "wide unicode" (why not?)

![image](https://cloud.githubusercontent.com/assets/13041/23109717/7c204d2a-f6e9-11e6-9ab4-8ab4b0709e73.png)

Another thing I'd like to add to your list: ensure chinese characters render correctly

rauchg commented:
More food for thought: how should `﷽` be rendered?

rauchg commented:
Personally, I like what Terminal.app does (similar to emoji handling). Take up one column, show overflow.

![image](https://cloud.githubusercontent.com/assets/13041/23109792/01d1a5ea-f6ea-11e6-9f24-67d1d7fe30f8.png)

iTerm takes up one column, but only shows partial overflow:

![image](https://cloud.githubusercontent.com/assets/13041/23109807/253281e4-f6ea-11e6-8dec-01b1ebece1ca.png)

tibotiber commented:
OK, that's what I had gathered about the encapsulation, just wanted to make sure there were no other reasons.

> That's not true for Emojis it seems? Or emojis are not "wide unicode" (why not?)

I don't understand. Emojis are wide unicode, and they do overflow on the next char in iTerm, don't they? This is what your screenshot shows with the heart overflowing from its span onto the next (behind the `a`). Maybe it's a vocabulary issue, I have the feeling we're saying the same thing.

Good job finding ﷽ 😂, I'll target full overflow, like in Terminal.app.

> Another thing I'd like to add to your list: ensure chinese characters render correctly

Which part exactly? The `wc-node` behaviour is done. Are you referring to the 2 chars that displayed as 3 mentioned in https://github.com/zeit/hyper/pull/1376#issuecomment-271876444 / https://github.com/zeit/hyper/pull/1376#issuecomment-272191734? Or anything else?

rauchg commented:
Just wanted to be 100% sure! Seems like we're on the same page @tibotiber.
Unicode nirvana (for _output) is within reach! 

Next up once we merge this PR is to fix backspace (the behavior of the ANSI escape that is) for surrogate pairs and caret position

rauchg commented:
Yep I was referring to the report by @BirkhoffLee

rauchg commented:
One super tricky edge case that'll be difficult to solve (with my current knowledge of CSS):

We want it to overflow, but if it's at the edge, we would want the parent container to trigger scrolling somehow.

No terminal that i know of seems to handle this edge case currently

![image](https://cloud.githubusercontent.com/assets/13041/23111166/be7e380c-f6f4-11e6-8e9c-1a6192e3c2ec.png)

rauchg commented:
Also not a priority for now considering our broken backspace, but thought I'd mention it

tibotiber commented:
Took a while to get back to this, sorry. I pushed new changes to try and fix the display of unicode in powerline prompts:
- removed existing unicode patching that was creating new containers for unicode chars. It was making it impossible for unicode to overflow as desired when there is a background color (e.g. powerline).
- instead I override hterm's `insertString` to get back the node in which the string has been added and I decompose its `textContent` when it contains unicode to insert child nodes wrapping unicode in individual spans. Doing so we still manage to get monospaced unicode but it can now overflow as long as the background color doesn't change.

This is partially working and I can't figure out why only "partially". It basically works for general unicode display such as `cat` files or `vtop`, it also work with file editors when re-rendering rows etc, but the node's content is flattened as if nothing happened in powerline prompts, except for the last node in a row.

I think it has something to do with the way hterm handles data decoding in `hterm.VT.prototype.parseUnknown_` but I can't manage to pinpoint exactly what. I could be wrong though and was hoping someone has an idea. I was originally not familiar with hterm's internals so I might miss something.

I left some debug logs that help to understand what the new `insertString` does.

tibotiber commented:
Wow, I almost forgot about this PR. Got back to it yesterday with fresh eyes and finally found what was missing. hterm's `insertString` was actually overriding my spans content to inject white spaces under some circumstances. Really nothing to do with the parser 😅.

So it is now where I wanted it to be, i.e. it encapsulate unicode nodes in a span within the current container, instead of creating a new container for each unicode node. This fixes the display of legacy powerline icons:
![old-powerline](https://cloud.githubusercontent.com/assets/5635553/26045254/f1a845aa-397b-11e7-9557-28500e6a729d.png)

wide chars (CJK) display is fixed, and unicode nodes are displayed with fixed width and full overflow as suggested by @rauchg:
![unicode](https://cloud.githubusercontent.com/assets/5635553/26045253/f1a33b96-397b-11e7-9fac-fce54e6386b4.png)

vtop and others are displaying well:
![vtop](https://cloud.githubusercontent.com/assets/5635553/26045252/f19f4676-397b-11e7-97cf-369c5c9e7805.png)

and 2-sided powerline prompts in tmux are now fixed:
![tmux](https://cloud.githubusercontent.com/assets/5635553/26045251/f19a8898-397b-11e7-9665-11a73a297ac2.png)


The only thing left is the issue where multiple CJK chars are rendered. I'll look into it, but we can also merge this PR first and when I fix this and I'll open another one. I'll let you guys review before I squash the commits (lots of trial and error, boo).

Cc @albinekb @rauchg

tibotiber commented:
Good news, I found the issue that caused multiple wide chars (e.g. CJK) to be rendered. It was a case where hyper's override of `lib.wc.substr` had `startIndex = undefined`. This PR is complete now 🎉. I fixed it by matching closer to the original code, despite using `runes` to be emoji-friendly.

![cjk](https://cloud.githubusercontent.com/assets/5635553/26055188/3ac000cc-39a2-11e7-8194-7d3402f29153.png)

albinekb commented:
Wow @tibotiber 😱

albinekb commented on lib/hterm.js:
```diff
@@ -388,6 +345,157 @@ hterm.Terminal.prototype.focusHyperCaret = function () {
+    return;
+  }
+  if (reverseOffset === 0) {
+    const nextSibling = cursorNode.nextSibling;
```
nit:
```js
const { nextSibling } = cursorNode;

tibotiber replied:
I'd rather keep it like the original `hterm` code to make diffs easier. Makes sense?

albinekb replied:
Absolutely! 👍

tibotiber commented:
@albinekb thank you for the review. Most of the voodoo magic is not mine actually, it's mainly the original `hterm` code that I patched as minimally as possible:
- replaced `var` by `const` or `let` to match local code style
- removed comments (it was extremely verbose and still available in the original code)
- made sure to call `wrapUnicode` in `insertString` and `splitNode_`

The `wrapUnicode` method is mine, so that might be a good place to focus on in reviews 😛.

albinekb commented:
After running `npm run dist` (macOS 10.12.4 (16E195)) I get 
![image](https://cloud.githubusercontent.com/assets/5027156/26270799/7438648e-3cfe-11e7-82c3-275f167ff46f.png) this error 🤔  Is this just me? @rauchg  @tibotiber

I'm gonna try master now as we speak..

rauchg commented:
I just got the same thing

rauchg commented:
Reverting

albinekb commented:
I'll have a look for a fix

tibotiber commented:
Hey, sorry about this, didn't try to `run dist`. I'll have a look as well.

tibotiber commented:
I'm not able to reproduce. I'm only getting a warning for an unknown OSC code.

<img width="588" alt="screen shot 2017-05-20 at 12 37 56 pm" src="https://cloud.githubusercontent.com/assets/5635553/26273011/41edfe52-3d59-11e7-94a0-cdc570b35a0b.png">

Might be coming from different prompts, what is yours like?

albinekb commented:
I use https://github.com/brandonweiss/pure.fish and i think rauch is using https://github.com/sindresorhus/pure @tibotiber

tibotiber commented:
Mmmmh, I'm on sindresorhus pure as well actually. I tried a few other prompts (antigen themes) and I'm getting no error. Not understanding this.

albinekb commented:
🤔 I'll diag & give you something to reproduce with tomorrow, need to get some 💤

mengzhuo commented:
Why don't we just let the webkit do this work instead of calculate the width of the char?

albinekb commented:
Sorry for not coming back yet, I'll get this sorted tomorrow 👼

mengzhuo commented:
@albinekb 

I use "Cousine for Powerline"

![2017-05-24 10-58-55](https://cloud.githubusercontent.com/assets/885662/26385131/114fdb60-4003-11e7-9baa-98ed5bcea1de.png)

After I remove all unicode-node style.

I think it's a "web-based" app, we should due with it more "web-way" by using more css instead of messing with DOM.

tibotiber commented:
@mengzhuo, the issue with this approach is that you break the rendering for unicode that is not monospaced (see [this comment](https://github.com/zeit/hyper/pull/1536#issuecomment-280972848)). You can observe a bad example by running `vtop` after removing unicode styling. More generally, terminals expect monospace chars and handle exceptions such as CJK with "wide-chars" which are doubled-spaced.

albinekb commented:
Sorry for the delay @tibotiber 

the problem is that not all nodes has `className`, for example text-nodes (see https://developer.mozilla.org/en-US/docs/Web/API/Node/nodeType)

I tried two things:

- adding a check in `hterm.Screen.prototype.wrapUnicode` to only handle element nodes (type 3), no errors but emojis get scrambled.
- adding a check for `node.className`, same problem

tibotiber commented:
@albinekb thanks for posting more details. I looked into it and here are some thoughts:
1. we seem to have an issue with cursor position. Basically the cursor moves with double spacing on single spaced unicode. Since `hterm.Terminal.prototype.eraseToRight` uses the cursor position to calculate how many chars should be deleted, it creates an issue with unicode deletion. I'm thinking this might be the source of the unknown char that appears when we delete emojis. Note: this issue exists in master so it is not new to this PR and could be tackled in another PR.
2. nice find on "the problem is that not all nodes has className", I completely missed it. I fixed this locally and will open a new PR soon since this one is now frozen.

tibotiber commented:
Quite honestly, `hterm` is horrible to patch because the code is extremely imperative and convoluted. Do you know of any other terminal emulator that would be more "react-ish"? That would be really cool! @rauchg?

LeonBlade commented:
I stumbled upon this PR and wanted to ask here before creating an issue first, so sorry in advance for posting this here.

By any chance is this rendering issue related to the problem produced in this screenshot: ![Terminal and Hyper](http://i.imgur.com/dBblSwW.png)

On the left is Terminal and on the right is Hyper.  I'm speaking of course about the box side which seems to warp as well as contain some overlap between the characters.

tibotiber commented:
@LeonBlade this will need to be checked once the follow up PR #1873 lands, but it should fix your issue I think.

uetchy commented:
I've tested it with Japanese and Korean characters on macOS El Capitan and there are some visual errors appeared.
There is an iTerm2 screen next to Hyper to clarify differences.
It looks like the width of Japanese characters is too narrow and Korean characters are weirdly broken.

![screenshot 2017-06-02 22 01 32](https://cloud.githubusercontent.com/assets/431808/26726731/25b185de-47df-11e7-8de9-6ef621ac3577.png)

ELLIOTTCABLE commented:
Hi! I don't necessarily have a horse in this race, besides the ones that the contributors above have (Pretty Unicode Interfaces [for english text] And Maybe Sometimes Emoji) — but I want to point out that this *heavily* affects users from other countries and cultures? This seems like a question that's ideal for a ‘reach out and ask users’ kind of moment.

If nothing else, I suspect we need to get a ~Unicode expert~ to come in and answer some questions about character ranges to treat as ‘wide’ (our implementation gives them double width, automatically consuming an extra column) versus ones to ignore, treat as single-wide, and leave for userspace handling (“What? This emoji looks funny? I guess I'd better spit out an extra space after it.” *aka* The Yarn Approach).

ELLIOTTCABLE commented:
An aside — copied [from a friend](https://twitter.com/FakeUnicode/status/886854168736092162), as an example of why the emoji in particular are A Hard Problem, and “display all private-use-area characters as two columns” isn't enough:

> 👨‍❤️‍💋‍👨 1f468-200d-2764-fe0f-200d-1f48b-200d-1f468 KISS: MAN, MAN
> 8 codepoints, 11 UCS-2 units, one Wide character.
> 
> — @‌FakeUnicode‬

sachin21 commented:
Great!! You saved my life. Thank you. I'll looking forward to be merged :+1:

sachin21 commented:
@tibotiber Great!! Thank you. I'll looking forward to be merged :+1: 

you saved my life :dango:

lucasqueiroz commented:
Seems like 1.4.x fixed this issue.
Vtop with 1.4.2 (PragmataPro Font):
![hyper_001](https://user-images.githubusercontent.com/29466590/30007536-673d35ec-90e7-11e7-9ee1-4bb661ce3652.png)
